### ISSUE TITLE
"Unhandled exception" when "cd ~"

### ISSUE THREAD

**Issue by marchandfab** [xonsh/xonsh#3474]
With an empty configuration file, when doing a simple `cd ~`, the following backtrace is displayed in the terminal:
```
Unhandled exception in event loop:
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/eventloop/utils.py", line 72, in schedule
    func()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 456, in redraw
    self._redraw()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 523, in _redraw
    self.context.run(run_in_context)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 509, in run_in_context
    self.renderer.render(self, self.layout)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/renderer.py", line 568, in render
    layout.container.preferred_height(size.columns, size.rows).preferred,
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 324, in preferred_height
    dimensions = [
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 325, in <listcomp>
    c.preferred_height(width, max_available_height) for c in self._all_children
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 792, in preferred_height
    return self.content.preferred_height(width, max_available_height)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 324, in preferred_height
    dimensions = [
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 325, in <listcomp>
    c.preferred_height(width, max_available_height) for c in self._all_children
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 2595, in preferred_height
    return self.content.preferred_height(width, max_available_height)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1620, in preferred_height
    return self._merge_dimensions(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1648, in _merge_dimensions
    preferred = get_preferred()
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/containers.py", line 1613, in preferred_content_height
    return self.content.preferred_height(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 639, in preferred_height
    content = self.create_content(width, height=1)  # Pass a dummy '1' as height.
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 787, in create_content
    cursor_position=translate_rowcol(
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 770, in translate_rowcol
    return Point(x=get_processed_line(row).source_to_display(col), y=row)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/layout/controls.py", line 716, in get_processed_line
    processed_line = transform(i, get_line(i))
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/lexers/pygments.py", line 319, in get_line
    for num, line in generator:
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/lexers/pygments.py", line 270, in create_line_generator
    yield from enumerate(split_lines(list(get_text_fragments())), start_lineno)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/lexers/pygments.py", line 268, in get_text_fragments
    yield _token_cache[t], v
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/lexers/pygments.py", line 141, in __missing__
    result = "class:" + pygments_token_to_classname(key)
  File "/usr/lib/python3.8/site-packages/prompt_toolkit/styles/pygments.py", line 66, in pygments_token_to_classname
    parts = ("pygments",) + token

Exception can only concatenate tuple (not "NoneType") to tuple
Press ENTER to continue...
```
Python: 3.8.2
xonsh: 0.9.14
Arch linux

**Comment by laloch** [comment]
Hi @marchandfab, thanks for reporting. This seems to be [prompt_toolkit3](https://github.com/prompt-toolkit/python-prompt-toolkit) upstream bug triggered by your pygments style configuration.
Can you reproduce the bug if you start `xonsh --no-rc`?
What is your Xonsh configuration?

**Comment by Briaoeuidhtns** [comment]
I'm on the same versions, and after an update I am getting the same exception. Happens when running with --no-rc also.

I tried rolling back to prompt_toolkit3=3.0.3 and pygments=2.5.2, but still same error. Rolling back xonsh from 0.9.14 to 0.9.13 with either version of prompt_toolkit did fix it for me.

Bisecting xonsh, the first commit that I can make fail is 6082b61f

**Comment by laloch** [comment]
Thanks for investigating the issue @Briaoeuidhtns. It seems to be related to `$LS_COLORS` environment variable once again.
Is the issue still present on current master/HEAD? If so, does the issue go away if you remove `$LS_COLORS` from the environment before you start Xonsh?

**Comment by Briaoeuidhtns** [comment]
It is still present at master/HEAD.

I actually didn't have $LS_COLORS set, but unset it just to be sure, no change.

Oddly enough, explicitly setting colors does seem to fix it, but only if done before launching. For instance, if I run `$LS_COLORS = $LS_COLORS.default_settings` from my working xonsh version before launching the one built off of current master, everything works as expected.

**Comment by melund** [comment]
@Briaoeuidhtns good digging :)  @bobhy Any idea what could be the going on here.

**Comment by bobhy** [comment]
Preliminarily, I can't repro with either ptk 3.0.3 or 3.0.4.  What version
of pygments, python etc?  Can anybody else confirm?

**Comment by Briaoeuidhtns** [comment]
prompt-toolkit 3.0.4, Pygments 2.6.1, python 3.8.2, running in a clean venv.
<details>
<summary>Everything</summary>
<pre>
appdirs==1.4.3
attrs==19.3.0
black==19.10b0
certifi==2019.11.28
cfgv==3.1.0
chardet==3.0.4
click==7.1.1
codecov==2.0.16
coverage==5.0.3
distlib==0.3.0
entrypoints==0.3
filelock==3.0.12
flake8==3.7.9
identify==1.4.11
idna==2.9
mccabe==0.6.1
more-itertools==8.2.0
nodeenv==1.3.5
packaging==20.3
pathspec==0.7.0
pluggy==0.13.1
pre-commit==2.1.1
prompt-toolkit==3.0.4
py==1.8.1
pycodestyle==2.5.0
pyflakes==2.1.1
Pygments==2.6.1
pyparsing==2.4.6
pytest==5.3.5
pytest-cov==2.8.1
pytest-flake8==1.0.4
pytest-timeout==1.3.4
PyYAML==5.3
regex==2020.2.20
requests==2.23.0
six==1.14.0
toml==0.10.0
typed-ast==1.4.1
urllib3==1.25.8
virtualenv==20.0.10
wcwidth==0.1.8
xonsh==0.9.14
</pre>
</details>

I can't reproduce on an arch docker with the same versions, but I'm not really sure what that leaves if @marchandfab is also having the same issue

**Comment by bobhy** [comment]
Thanks for the additional info.   I've got a solid repro now, you had it right in the first report.  My previous checkin can return `None` for a token when trying to colorize a file system object (directory or file),  pygments doesn't like that.  Working on a fix now...

FYI, simple repro with `xonsh 0.9.14' and `prompt_toolkit 3.0.3`, does not depend on empty `xonshrc`.but does depend on timing:

```python
# (invoke xonsh with normal .xonshrc)
> $LS_COLORS={}
> xonsh --no-rc
> ls work # if 'work' is a subdirectory name, will crash when type 'k'
```
It will not crash (for me) if I swap 2nd and 3rd commands, and does colorize the file name with a value not in $LS_COLORS (by definition), so there's a second bug lurking in there...

**Comment by bobhy** [comment]
Fix available in #3477; works interactively (e.g, with above repro), but new unit tests failing apparently due to events not firing.  Any suggestions about the latter appreciated...
